typed-racket-lib/typed-racket/base-env/unit-prims.rkt

```diff
@@ -0,0 +1,767 @@
+  (define-syntax-class unit-imports
+    #:literals (import)
+    (pattern (import sig:sig-spec ...)
+             #:with names #'(sig.sig-name ...)
+             #:attr renamers (attribute sig.rename)))
+
+  (define-syntax-class unit-exports
+    #:literals (export)
+    (pattern (export sig:sig-spec ...)
+             #:with names #'(sig.sig-name ...)
+             #:attr renamers (attribute sig.rename)))
+  
+  (define-syntax-class sig-spec
+    #:literals (prefix rename)
+    (pattern sig-id:id
+             #:attr rename (lambda (id) id)
+             #:with sig-name #'sig-id)
+    (pattern (prefix p:id sig:sig-spec)
+             #:attr rename (lambda (id) (format-id #'sig.sig-name
+                                                 "~a~a"
+                                                 #'p
+                                                 ((attribute sig.rename) id)))
+             #:with sig-name #'sig.sig-name)
+    (pattern (rename sig:sig-spec (new:id old:id) ...)
+             #:attr rename 
+             (lambda (id)
+               (define mapping  (map cons  
+                                     (syntax->list #'(old ...))
+                                     (syntax->list #'(new ...))))
+               (define (lookup id) 
+                 (define lu (member id mapping 
+                                    (lambda (x y) 
+                                      (free-identifier=? x (car y)))))
+                 (and lu (cdar lu)))
+
+               (define rn ((attribute sig.rename) id))
+               (or (lookup rn) rn))
+             #:with sig-name #'sig.sig-name))
+  
+  ;; need lexical signature vars to error with duplicate type annotations
+  (define (signatures-vars stx)
+    (define (signature-vars sig-id)
+      (let-values ([(_0 vars _2 _3)
+                    (signature-members sig-id sig-id)])
+        vars))
+    (apply append (map signature-vars (syntax->list stx))))
+  
+  ;; extract vars from a signature with the correct syntax marks
+  ;; TODO: this is probably not general enough and will need to be modified
+  (define (get-signature-vars sig-id)
+    (define-values (_0 vars _2 _3)
+      ;; TODO: give better argument for error-stx
+      (signature-members sig-id sig-id))
+    (map
+     syntax-local-introduce
+     vars)))
+
+
+;; Abstraction for creating trampolining macros
+(begin-for-syntax
+  (define-syntax-class (begin-form name [arg #'()])
+    #:literals (begin)
+    (pattern (begin e ...)
+             #:with trampoline-form
+             #`(#,name #,@arg e ...)))
+  (define-syntax-class (name-form name)
+    (pattern (_ e ...)
+             #:with trampoline-form
+             #`(begin (#,name e) ...)))
+  (define-splicing-syntax-class (rest-form name arg)
+    (pattern (~seq) 
+             #:with trampoline-form
+             #`(begin))
+    (pattern (~seq e1 e2 ...)
+             #:with trampoline-form
+             #`(begin (#,name #,@arg e1)
+                      (#,name #,@arg e2) ...))))
+
```

Comment:
Can you provide a comment explaining how this abstraction works / is intended to be used?